### ISSUE TITLE
[Explore]: Prevent duplicate records (both name and GUID)

### ISSUE THREAD

**Issue by matt-royal** [cloudfoundry/cf-k8s-controllers#383]
### Background

We are currently taking several different approaches to avoid duplicate records:
- We guard against multiple apps in a space with the same name using a [validating web hook](https://github.com/cloudfoundry/cf-k8s-controllers/blob/main/controllers/webhooks/workloads/cfapp_validation.go)
- Several of our repo fetch methods in the api codebase [will error if they find duplicate results](https://github.com/cloudfoundry/cf-k8s-controllers/blob/203306701a4825584a514138bcd8cf73c877d7e4/api/repositories/app_repository.go#L440-L452)
- We have a ["leases" proposal](https://docs.google.com/document/d/1fYIulNhfXlp_y2Ah5EtoKDDX83s6V-WRu162xpziLZ8/edit?usp=sharing) to enforce uniqueness, but I don't believe it's implemented for any resources yet. It's also fairly heavy weight.
- There's a [proposal to use CF names as K8s names](https://docs.google.com/document/d/1swH9ZJgxluo1GIJnqCn54gNLosyd-aJqgynzHJkGH1I/edit), which would ensure uniqueness of name within a namespace. However, this has problems, as per the comments in the proposal

Generally at creation time we're not ensuring uniqueness of GUIDs (`metadata.name`) across namespaces, and of names (`spec.name`) within namespace.

### Acceptance Criteria

Explore the current solutions to this problem and synthesize a few options for how we can enforce global uniqueness of GUIDs and uniqueness of names within a space. We may use several different solutions at once depending on the type of resource, how dangerous it is to have duplicates, and how expensive and error prone the solution is.

### Number of days this exploration is expected to take. This is what is “pointed”.

3 days.

### Dev Notes

- Ideally we will prevent the resources from being created in the k8s API. Failing that, we should transition the duplicate record into an error state in the k8s API and ignore it in the CF API shim.

- The solution should prevent duplicates when using kubectl directly, not just when using the CF API

- Consider Open Policy Agent

**Comment by kieron-dev** [comment]
Just to clarify we have the leases solution, wrapped in [NameRegistry](https://github.com/cloudfoundry/cf-k8s-controllers/blob/main/controllers/coordination/name_registry.go), for orgs, spaces and apps currently.

**Comment by kieron-dev** [comment]
Here's a link to a spike done on OPA for the authorization work: https://github.com/eirini-playground/auth-explore/tree/master/opa-gatekeeper, with some words [here](https://docs.google.com/document/d/1ptkqMJjOPrjPz7T2yo1FaMeTA3T5klgBJgAcbxARgTs/edit#heading=h.eadz3lxsgxg0).

**Comment by julian-hj** [comment]
**TL;DR**: We think the lease mechanism looks like a good approach; if we can find a way to share code across validating web hooks instead of copy/pasting, that would be better.

@acosta11 & I walked through all of the various entity types that it is possible to create today with the CF cli and double checked against the current state of our product. As it stands today, orgs, spaces, and apps are all protected with the lease/NameRegsitry mechanism above.

The only types that are implemented, but not currently protected with the lease mechanism are CFRoute and CFDomain. These types currently have no checks for duplicate names.

We did a quick POC to see how difficult it would be add a web hook to detect duplicate routes. It's in [this branch](https://github.com/cloudfoundry/cf-k8s-controllers/tree/issues/383-duplicate-names)

This implementation took about 2ish hours to get working, so not too bad. (we skipped tests of any kind, so maybe it would have been a day to make it perfect)

We copy-pasted a lot of code from the CFApp webhook. Maybe we want to consider some mechanism to make the checking logic generic? We would need some interface to find a "lock name" string for each type, and to produce error messages for the various error cases, but otherwise the implementation is very standard.

**Comment by kieron-dev** [comment]
I like the sound of a generic, parameterisable webhook that can act on all the types.

The org/space logic is slightly different at the moment, as you have to look at a label on a subnamespaceanchor to tell whether it's an org or a space. But once we have CRDs for orgs and spaces that distinction vanishes.

**Comment by matt-royal** [comment]
Thank you, @kieron-dev, for the info and for weighing in. The leases solution is much farther along than I'd realized and it's fantastic that we a generic solution on top of it is feasible!

Thanks for the work and the write-up, @julian-hj and @acosta11. I'll write a follow up story to implement the missing uniqueness constraints, with a dev note about generalizing it.

**Comment by matt-royal** [comment]
@julian-hj @acosta11 I just realized that you didn't discuss GUID uniqueness in your write-up. Is there anything currently insuring that we don't have duplicate GUIDs across spaces?

**Comment by matt-royal** [comment]
We talked via Slack, and they decided not to focus on GUID uniqueness since it isn't an issue when using the API shim rather than kubectl directly. I'm going to spin off that exploration into a separate, ice-boxed story.

Looking at the resources in previous comments, it looks like you could use Leases to ensure resource GUID uniqueness across spaces by making the Leases in the root namespace. This would prevent situations where `/v3/apps/some-app-guid` matches multiple resources.
